### ISSUE TITLE
Restore CouponCode After Order Cancellation

### ISSUE THREAD

**Issue by agorbulin** [magento/magento2#19545]
### Description (*)
**Problem:** when customer places order using personal (one time) coupon code
and order will be cancelled by PSP or manually, customer can not use it again.
**Magento ver.:** Reproduced on 2.2.* and 2.3.*

### Expected behavior (*)
Coupon code should be restored for re-use, if order cancelled

### Benefits
We have request from our Commerce customer for this feature and I've found request about this feature https://github.com/OpenMage/magento-lts/issues/167 , which not resolved

### Additional information
**STR:**
- Create cart rule with coupon code (e.g. '1euro')
- Set for rule 'Uses per Customer' = 1
- Login as customer
- Place product to the cart and apply coupon code (applied successfully)
- Place order
- Cancel order (manually in BO or by a PSP)
- Place some product to the card again and apply coupon code one more time

**Expected result:**
Coupon code applied

**Actual result:**
Coupon code is not applied (error like this: Discount code "1euro" is not valid.)

**Comment by magento-engcom-team** [comment]
Hi @agorbulin. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me $VERSION instance` 

where `$VERSION` is [version tags](https://github.com/magento/magento2/tags) (starting from 2.2.0+) or develop branches (for example: 2.3-develop). 
For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@agorbulin do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [x] yes
- [ ] no

**Comment by kunj1988** [comment]
I just reproduce the issue in Magento 2.3

**Comment by orlangur** [comment]
@magento-engcom-team please evaluate such system behavior change. Should it be switchable preserving old behavior by default?

Removing labels so that issue can be processed properly.

**Comment by kunj1988** [comment]
@magento-engcom-team , can you assign this to me?

**Comment by orlangur** [comment]
@kunj1988 what do you mean? There is a PR with fix provided already.

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-nazar. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [x] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [x] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [x] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [x] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [x] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [x] 6. Add label `Issue: Confirmed` once verification is complete. 

- [x] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by tkacheva** [comment]
@agorbulin Thanks for the contribution. If customer didn't get the coupon benefit due to order cancellation, coupon needs to be active unless the promotion has end date that is already due.

**Comment by agorbulin** [comment]
Thank you @tkacheva for the clarification.
I will proceed with issue and add verification of  'salesrule end date'.

**Comment by magento-engcom-team** [comment]
@engcom-backlog-nazar Thank you for verifying the issue. Based on the provided information internal tickets `MAGETWO-97035`, `MAGETWO-97036` were created

**Comment by desaibharat** [comment]
I will work on this issue #mm19in

**Comment by magento-engcom-team** [comment]
@desaibharat thank you for joining. Please accept team invitation [here](https://github.com/orgs/magento/invitation) and self-assign the issue.

**Comment by magento-engcom-team** [comment]
Hi @desaibharat. Thank you for working on this issue. 
Looks like this issue is already verified and confirmed. But if your want to validate it one more time, please, go though the following instruction:
- [ ] 1. Add/Edit `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.

- [ ] 2. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 3. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 4. If the issue is not relevant or is not reproducible any more, feel free to close it.

**Comment by deepaksnair123** [comment]
@agorbulin , @magento-engcom-team any update on this issue??

**Comment by engcom-Bravo** [comment]
Hello @agorbulin 
Thank You for contribution and collaboration!

We are closing this issue due to _branch `2.2-develop` was closed and Magento no longer accepts pull requests and issues to the v2.2 release lines to focus all development efforts on v2.3._
 See [Accepted pull requests and ported code](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#rules) for more details. Also we are not able to reproduce this issue on the lates `2.3-develop` branch by provided steps that takes place because the issue most likely has been fixed.

If You still manage to reproduce this issue on `2.3-develop` please create a new Issue with all required details according to [Issue reporting guidelines](https://github.com/magento/magento2/wiki/Issue-reporting-guidelines)
